### PR TITLE
[6.3] Restore activation key and content hosts subscriptions (BZ:1296978)

### PR DESCRIPTION
**Pull request by ldjebran**
cover: https://bugzilla.redhat.com/show_bug.cgi?id=1296978
```console
(sat-6.3.0) dlezz@elysion:~/projects/robottelo-fork$ pytest -v tests/foreman/cli/test_subscription.py::SubscriptionTestCase::test_positive_restore_ak_and_content_hosts_subscriptions
================================================= test session starts ==================================================
platform linux2 -- Python 2.7.13, pytest-3.1.3, py-1.4.34, pluggy-0.4.0 -- /home/dlezz/.pyenv/versions/sat-6.3.0/bin/python2.7
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: xdist-1.20.0, services-1.2.1, mock-1.6.2, forked-0.2, cov-2.5.1
collected 1 item 
2017-10-17 10:51:47 - conftest - DEBUG - Deselect of WONTFIX BZs is disabled in settings


tests/foreman/cli/test_subscription.py::SubscriptionTestCase::test_positive_restore_ak_and_content_hosts_subscriptions PASSED

============================================== 1 passed in 656.55 seconds ==============================================
```

### CODE REVIEW

File: tests/foreman/cli/test_subscription.py [hunk 1]
```diff
@@ -189,3 +199,169 @@ def test_negative_manifest_refresh(self):
...
             })
+
+    @tier3
+    def test_positive_restore_ak_and_content_hosts_subscriptions(self):
+        """Restore activation key and content hosts subscriptions
+
+        :id: a44fdeda-9c8c-4316-85b4-a9b6b9f1ffdb
+
+        :steps:
+            1. Setup activation key , lifecycle environment and content view
```
Comment (oshtaier) [inline tests/foreman/cli/test_subscription.py]: Where did you get these steps? I read BZ information and it has some deviations from your steps

Comment (oshtaier) [reply tests/foreman/cli/test_subscription.py]: Especially that csv was edited to another host/capsules before applying

Comment (ldjebran) [reply tests/foreman/cli/test_subscription.py]: @oshtaier to make the exact BZ steps we have to be able to get an other manifest with the same subscriptions but with different agreement number, that is not realizable with the current infra implementation (we need muti-manifests s to be supported), the current  coverage implement that by deleting completely the manifest and re-import all data, if you insist on the exact implementation we have to make this PR in do not merge state (marked as blocked ) and wait that PR implementation (multi-manifest) to be merged.

Comment (oshtaier) [reply tests/foreman/cli/test_subscription.py]: Hmm. Coverage that you provided in that PR is also needed, but still I am not sure that it is actually cover mentioned BZ

Comment (abalakh) [reply tests/foreman/cli/test_subscription.py]: +1 to @oshtaier . The steps are looking different to me especially without changing subscription id in csv.
We need @JacobCallahan approval whether these steps fully cover the BZ, and if not - corresponding framework expansion is needed.

Comment (JacobCallahan) [reply tests/foreman/cli/test_subscription.py]: @abalakh @ldjebran There is one primary case that should be covered in the BZ, which the current test does not cover. The ideal steps for the bug would be:
1. Import a manifest, with a couple of subscriptions.
2. Register at least one host (preferably more).
3. Attach the same subscription to all registered hosts.
4. Export the content hosts to a csv file.
5. Modify the csv file to change the hosts from their current subscription to a different one.
6. Import the csv file
Result: The hosts should no longer have the old subscription, and now have the new one.

There is also an additional test case that could be written for what I did in comment 36. This is pretty similar to the steps above, but does not require hosts to be registered.
1. Upload a manifest with a couple of subscriptions.
2. Create an activation key, aligning one of the subscriptions to it.
3. Export the activation key to a csv file.
4. Modify the csv file to change the activation key's subscriptions from the previous sub to a different one.
5. Import the csv file.
Result: The activation key should not have the old subscription, and should have the new one.

With that said, this is still a very valid test that we should have in the code, just not aligned to this bug.

**Comment by ldjebran** [comment]
PUT dot merge to make the changes

**Comment by ldjebran** [comment]
@abalakh @JacobCallahan @oshtaier comments addressed
```
(sat-6.3.0) dlezz@elysion:~/projects/robottelo-fork$ pytest tests/foreman/cli/test_subscription.py -v -k "_restore_"
=================================================================================================== test session starts ====================================================================================================
platform linux2 -- Python 2.7.13, pytest-3.2.3, py-1.4.34, pluggy-0.4.0 -- /home/dlezz/.pyenv/versions/2.7.13/envs/sat-6.3.0/bin/python
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: xdist-1.20.1, services-1.2.1, mock-1.6.3, forked-0.2, cov-2.5.1, fauxfactory-1.0.1
collected 9 items                                                                                                                                                                                                           
2017-11-21 10:27:16 - conftest - DEBUG - BZ deselect is disabled in settings


tests/foreman/cli/test_subscription.py::SubscriptionTestCase::test_positive_restore_ak_and_content_hosts_subscriptions PASSED
tests/foreman/cli/test_subscription.py::SubscriptionTestCase::test_positive_restore_ak_with_modified_subscription PASSED
tests/foreman/cli/test_subscription.py::SubscriptionTestCase::test_positive_restore_content_hosts_with_modified_subscription PASSED

==================================================================================================== 6 tests deselected ====================================================================================================
============================================================
```

**Comment by ldjebran** [comment]
@renzon have added two more tests to this PR please revisit

**Comment by renzon** [comment]
ACK

File: tests/foreman/cli/test_subscription.py [hunk 2]
```diff
@@ -189,3 +241,475 @@ def test_negative_manifest_refresh(self):
...
+            download_file(
+                chs_export_remote_file_path, chs_export_local_file_path)
+            # modify the content hosts subscription
+            field_names, csv_data = self._read_csv_file(
+                chs_export_local_file_path)
+            # each client is represented by one row of data
+            self.assertEqual(len(csv_data), 2)
+            for row_data in csv_data:
+                # The subscription is saved in the following format:
+                # """<quantity>|<sku>|<name>|<contract>|<account>"""
```
Comment (abalakh) [inline tests/foreman/cli/test_subscription.py]: bonus karma for such comments, they're very helpful for understanding the code 👍

**Pull request merged**